### DISCUSSION TITLE
std::path::Path::new().exists() with python Pathlib.path() works in linux and windows but not macos

### DISCUSSION THREAD

**Discussion by baseplate-admin** [PyO3/pyo3#4137] [Questions]
Hi i have this following code

rust:
```rust
    let svg_string: String;

    if std::path::Path::new(&svg).exists() {
        let mut svg_data = std::fs::read(svg)
            .map_err(|_| "failed to open the provided file")
            .unwrap();
        if svg_data.starts_with(&[0x1f, 0x8b]) {
            svg_data = resvg::usvg::decompress_svgz(&svg_data)
                .map_err(|e| e.to_string())
                .unwrap();
        };
        svg_string = std::str::from_utf8(&svg_data).unwrap().to_owned();
    } else {
        svg_string = svg;
    }

```

```python


def test_path():
    path = os.path.join(BASE_DIR, "acid.svg")
    base = resvg_py.svg_to_base64()
    print(path)
    assert base == svg_output


def test_gzip_path():
    path = os.path.join(BASE_DIR, "acid.svg.gz")
    base = resvg_py.svg_to_base64()
    print(path)

    assert base == svg_output
```

This fails in macos.

Here is the  log : https://github.com/baseplate-admin/resvg-py/actions/runs/8889901090/job/24409004312
Relevant Source :
* Rust : https://github.com/baseplate-admin/resvg-py/blob/4a89a841138d3297986892e6418c777fb068c140/src/rust/lib.rs#L164-L178 
* Python : https://github.com/baseplate-admin/resvg-py/blob/e981e211fccd43cf0581d870e0fdfb3187667023/tests/test_path.py#L1-L22

**Comment by birkenfeld** [comment]
That doesn't seem related to Path. In any case, `os.path.join` will return a string even when called with a `Path`.

The relevant panic message here is: "unknown token at 1:1" which seems to come from https://github.com/RazrFalcon/roxmltree/blob/master/src/parse.rs#L297

and that probably originates from https://github.com/baseplate-admin/resvg-py/blob/4a89a841138d3297986892e6418c777fb068c140/src/rust/lib.rs#L114

So in some way, parsing the SVG fails on MacOS...

**Reply by baseplate-admin** [reply]
Hmm, very weird. 

Because if i modify this [code](https://github.com/baseplate-admin/resvg-py/blob/4a89a841138d3297986892e6418c777fb068c140/src/rust/lib.rs#L164-L178):

```rust
   let svg_string: String;

    if std::path::Path::new(&svg).exists() {
        let mut svg_data = std::fs::read(svg)
            .map_err(|_| "failed to open the provided file")
            .unwrap();
        if svg_data.starts_with(&[0x1f, 0x8b]) {
            svg_data = resvg::usvg::decompress_svgz(&svg_data)
                .map_err(|e| e.to_string())
                .unwrap();
        };
        svg_string = std::str::from_utf8(&svg_data).unwrap().to_owned();
    } else {
        svg_string = svg;
    }
```

and remove the file read logic [code](https://github.com/baseplate-admin/resvg-py/blob/64f425eae0478584a3ebdadf3974602482a0559b/src/rust/lib.rs#L164): 

```rust
    let svg_string: String = svg;
````


The tests are passing  (well file read doesn't work) : https://github.com/baseplate-admin/resvg-py/actions/runs/8893280016/job/24419112423


---


But since i dont have a macbook, i am not able to debug this issue locally.

---

> So in some way, parsing the SVG fails on MacOS...

This is interesting, same code runs in `windows` / `linux` but fails in macos ...

**Reply by birkenfeld** [reply]
Ohh I just saw what you're doing there. (My advice: don't. An argument that can be both a path or content is a recipe for disaster.)

**Reply by baseplate-admin** [reply]
> My advice: don't. An argument that can be both a path or content is a recipe for disaster.

i agree with you. But this is a requirement for a project.

----

i never thought this would be painful to implement

**Reply by baseplate-admin** [reply]
Hi @birkenfeld,

i have found another issue in macos.

https://github.com/PyO3/pyo3/issues/4161
